business/authentication/openshift_auth_controller.go

```diff
@@ -0,0 +1,191 @@
+	bs, err := o.businessInstantiator(&api.AuthInfo{Token: ""})
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving the OAuth package (getting business layer): %w", err)
+	}
+
+	user, err := bs.OpenshiftOAuth.GetUserInfo(token)
+	if err != nil {
+		o.SessionStore.TerminateSession(r, w)
+		return nil, &AuthenticationFailureError{
+			Reason:     "Token is not valid or is expired.",
+			Detail:     err,
+			HttpStatus: http.StatusUnauthorized,
+		}
+	}
+
+	err = o.SessionStore.CreateSession(r, w, config.AuthStrategyOpenshift, expiresOn, openshiftSessionPayload{Token: token})
+	if err != nil {
+		return nil, err
+	}
+
+	return &UserSessionData{
+		ExpiresOn: expiresOn,
+		Username:  user.Metadata.Name,
+		Token:     token,
+	}, nil
+}
+
+// ValidateSession restores a session previously created by the Authenticate function. The user token (access_token)
+// is revalidated by re-fetching user info from the cluster, to ensure that the token hasn't been revoked.
+// If the session is still valid, a populated UserSessionData is returned. Otherwise, nil is returned.
+func (o openshiftAuthController) ValidateSession(r *http.Request, w http.ResponseWriter) (*UserSessionData, error) {
+	sPayload := openshiftSessionPayload{}
+	sData, err := o.SessionStore.ReadSession(r, w, &sPayload)
+	if err != nil {
+		log.Warningf("Could not read the openshift session: %v", err)
+		return nil, nil
+	}
+	if sData == nil {
+		return nil, nil
+	}
+
+	// The Openshift token must be present
+	if len(sPayload.Token) == 0 {
+		log.Warning("Session is invalid: the Openshift token is absent")
+		return nil, nil
+	}
+
+	bs, err := o.businessInstantiator(&api.AuthInfo{Token: sPayload.Token})
+	if err != nil {
+		log.Warningf("Could not get the business layer!: %v", err)
+		return nil, fmt.Errorf("could not get the business layer: %w", err)
+	}
+
+	user, err := bs.OpenshiftOAuth.GetUserInfo(sPayload.Token)
+	if err == nil {
+		// Internal header used to propagate the subject of the request for audit purposes
+		r.Header.Add("Kiali-User", user.Metadata.Name)
+		return &UserSessionData{
+			ExpiresOn: sData.ExpiresOn,
+			Username:  user.Metadata.Name,
+			Token:     sPayload.Token,
+		}, nil
+	}
+
+	log.Warningf("Token error: %v", err)
+	return nil, nil
+}
+
+// TerminateSession session created by the Authenticate function.
+// To properly clean the session, the OpenShift access_token is revoked/deleted by making a call
+// to the relevant OpenShift API. If this process fails, the session is not cleared and an error
+// is returned.
+// The cleanup is done assuming the access_token was issued to be used only in Kiali.
+func (o openshiftAuthController) TerminateSession(r *http.Request, w http.ResponseWriter) error {
+	sPayload := openshiftSessionPayload{}
+	sData, err := o.SessionStore.ReadSession(r, w, &sPayload)
+	if err != nil {
+		return TerminateSessionError{
+			Message:    fmt.Sprintf("Apparently, there is no active openshift session: %v", err),
```

Comment:
> It makes it sound like the server is surprised which doesn't instill confidence or could be confusing. Although it does add a nice human touch.

Haha... Yes! The server should be surprised! Now, I am tempted to add: _Normally, the UI would kick you to the login screen as soon as the session expires. But somehow you managed to stay at the console and were able to click "Logout" on it... how it could be? What are you doing?_ :-P

But seriously, since @jmazzitelli has already commented, let's ask about his opinion. English is your first language, so if you both agree that "Apparently" shouldn't be here, I will remove it.

Reply:
I would just say something like `There is no active openshift session`. That's good enough.

You could add some commentary, but I'm not sure it is needed. Perhaps, `There is no active openshift session. Your session has either timed out or you have been logged out.` (if that is even correct :)

Reply:
OK, I'm keeping it short. I removed the "Apparently".